### ISSUE TITLE
Files cache not effective?

### ISSUE THREAD

**Issue by sourcejedi** [borgbackup/borg#3357]
I'm not managing to get the benefit of the files cache in my testing.

> Backup speed is increased by not reprocessing files that are already part of existing archives and weren’t modified. The detection of unmodified files is done by comparing multiple file metadata values with previous values kept in the files cache.

I used v1.1.2, downloaded as the single big binary for x86_64.  Same results on Debian 9 and on Fedora 27.

Same results also if I rm BORG_REPO, ~/.cache/borg, ~/.config/borg, and then always pass `--files-cache=mtime,inode,size` to `borg create`.

export BORG_REPO=~/borgtest
mkdir testdata
cd testdata
dd if=/dev/zero of=testfile bs=1M count=240

borg init -e none
borg create ::{now} ~/testdata

strace -t -f -e read,fadvise64 borg create ::{now} ~/testdata  # now re-run create, and look at the IO

```
...
[pid 19838] 11:38:49 read(7, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 8388608) = 8388608
[pid 19838] 11:38:49 fadvise64(7, 8388608, 8388608, POSIX_FADV_DONTNEED) = 0
...
[pid 19838] 11:38:53 read(7, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 8388608) = 8388608
[pid 19838] 11:38:53 fadvise64(7, 243269632, 8388608, POSIX_FADV_DONTNEED) = 0
[pid 19838] 11:38:53 read(7, "", 8388608) = 0
[pid 19838] 11:38:53 fadvise64(7, 251658240, 0, POSIX_FADV_DONTNEED) = 0
```

**Comment by sourcejedi** [comment]
Same results with the 1.0.11 release, except that there is no --files-cache option, but AIUI it uses mtime,inodes,size already.  (Tested on Fedora 27).

My filesystems are ext4, there shouldn't be anything exciting about them.

**Comment by sourcejedi** [comment]
However when I did a test on real data (20G of photos), borg 1.1.2 on Debian 9, it worked as expected:

Initial create:

real	14m55.198s
user	5m4.044s
sys	0m47.976s

Second create:

real	0m6.574s
user	0m5.404s
sys	0m0.340s

So this makes me think of the other bug I reported when testing with gigabytes of zeros.  But the previous one affected the repo, not the files cache, it needed more data than this.  Also it was more an issue with the statistics reported, whereas this affects the (performance) functioning of backups.

**Comment by ThomasWaldmann** [comment]
See there:
http://borgbackup.readthedocs.io/en/stable/faq.html#i-am-seeing-a-added-status-for-an-unchanged-file

**Comment by sourcejedi** [comment]
Right.  I found that via another issue, and was in the process of whining about it.  (The real data test was very impressive, thanks!).

I think it's a natural and even fair evaluation test to run.  I can see it's really hard to improve though, once you have the aim to avoid this data-loss window.

> This is expected: it is to avoid data loss with files that are backed up from a snapshot and that are immediately changed after the snapshot (but within mtime granularity time, so the mtime would not change). Without the code that removes these files from the files cache, the change that happened right after the snapshot would not be contained in the next backup as Borg would think the file is unchanged.

Strawman brainstorming:

* I'd be happy with just sleeping 2 seconds at the end (for FAT, the worst case).  The problem is I'm not sure how to document that any better than the current situation.
* If sleeping 2 seconds at the end takes more than 10% of total run time, log a warning to inform the user?
* Alternatively, if we spent more than 2 seconds total chunking the files which had the latest mtime, then we're already safe and we don't need to worry about anything.  The files cache entries can be left alone.
* if we spent less than 2 seconds chunking those files, maybe forcing them to be re-chunked on the next run is not too bad.
* An evaluation test where a full chunking takes less than 2 seconds could give misleading timings given the above.  I think that's fine because such a test would not be a very good one.

**Comment by ThomasWaldmann** [comment]
Borg can't do anything about this. Snapshotting is done outside of borg. It can happen just before or long before, borg does not know.

But a known workaround (if you have big file(s) with latest timestamp) is to touch a small file within the input data, so that file gets the newest timestamp. If you do a snapshot, do it after touching.

**Comment by sourcejedi** [comment]
Ah.  The situation is different if you're relying on atomic snapshots (v.s. assuming mutual exclusion).  I had not considered that.

**Comment by sourcejedi** [comment]
This bugs me, the details are unfamiliar and I had difficulty remembering them. (And I happen to see the `bup` README again, which gives me the wrong impression). I'd like to write one more attempt at justifying `borg`'s behaviour.

borg does this to avoid a problem in case you took two snapshots e.g. in the same millisecond.  That leaves plenty of time between them for changes to be made.  But the modified times on the filesystem might not show a difference if it only records them [to the nearest millisecond](https://github.com/zfsonlinux/zfs/issues/1663#issuecomment-22833725).  If you archived the snapshots into borg in sequence, mounting them on the same point point to use the files cache - and try to avoid spending time chunking unchanged files - you lose.

borg is avoiding the problem that the later changes wouldn't be backed up.  *Including in subsequent backups*, unless the file was modified again.  borg does not want be responsible for losing data. Also,  `borg` doesn't want to turn a consistent atomic snapshot into an inconsistent one.

E.g. this could happen if you script regular filesystem snapshots, and independently schedule a borg script which finds and backs up all the new filesystem snapshots.  Maybe you manually requested a filesystem snapshot in the same millisecond that the scheduled snapshot was created.  It's not really possible for `bup` to rule this out.

It's not really possible for `borg` to rule out these conditions.  The workaround it uses is elegant, and in most cir
